### ISSUE TITLE
Android not receiving notifications when in background

### ISSUE THREAD

**Issue by globules-io** [dpa99c/cordova-plugin-firebasex#532]
Hello,
Everything works except one thing. 
- App is in the background
- Notification is received on the device
- If user does not tap the notification but instead switch back to the app
- Notification is not passed to the app

We tested on Android 6. To make sure it's not because the app is dozed, we did 2 tests.
- start the app
- put it in the background for 1 sec
- send notification from Firebase console
- device receives notification
- switch back to app (in total app has been in background 2 seconds or so) without tapping
- nothing passed to app

We also added cordova-background-mode plugin to make sure the app is running in the background and we have the same result. 

How do you handle this case ? Thank you.

**Comment by dpa99c** [comment]
On Android, when a notification message arrives on the device and the app is in the background/not running, the notification payload is delivered a centralised Android service (not the app) in order to display it as a system notification.
The notification payload is only delivered to the app if the user taps the system notification; otherwise it is not delivered.

To ensure the app always receives the notification payload, you should duplicate the message payload in the `notification` section of the message into the `data` section since the `data` payload is always delivered to the app, regardless of whether the user taps a notification. 
If the app is not running/in the background when a message containing a `data` payload arrives, the payload will be delivered in the extras bundle of the launch Intent of the app.
This plugin handles this when its [`pluginIntialize` method](https://github.com/dpa99c/cordova-plugin-firebasex/blob/master/src/android/FirebasePlugin.java#L177) is called during app startup - messages found in the extras bundle are added to the notification stack to be delivered along with the next message that arrives.


Please see also:
- [Receive messages in an Android app](https://firebase.google.com/docs/cloud-messaging/android/receive)
- [Firebase Notifications in Background & Foreground in Android](https://wajahatkarim.com/2018/05/firebase-notifications-in-background--foreground-in-android/)

**Comment by globules-io** [comment]
Thank you for your reply.

"If the app is not running/in the background when a message containing a data payload arrives, the payload will be delivered in the extras bundle of the launch Intent of the app."

We are using the data payload and it is never delivered here. It is only delivered if the app is in the foreground, or the notification has been tapped.

window.FirebasePlugin.onMessageReceived never receives anything in the case we described. Something we are missing ?

**Comment by globules-io** [comment]
Or do you mean we need to send twice, one notification with a notification title & body without data and one with only data and no notification title & body ?

**Comment by dpa99c** [comment]
This part of the plugin code is [directly inherited from `cordova-plugin-firebase`](https://github.com/arnesson/cordova-plugin-firebase/blob/master/src/android/FirebasePlugin.java#L80) so it's entirely possible there is a bug in it - it is not functionality I have needed (to detect notifications being delivered when they are not tapped by the user) so I haven't encountered this problem myself.
I would suggest debugging in Android Studio and placing a breakpoint [here](https://github.com/arnesson/cordova-plugin-firebase/blob/master/src/android/FirebasePlugin.java#L80) and see what the extras bundle contains at app launch after a message is received while in the background/not running.

**Comment by globules-io** [comment]
We added 

`Log.v(TAG, "______________________________Notification extras : " + extras.toString());`

before

`if (extras != null && extras.size() > 1) {`

just to see what's up. Here's the log:

`10-06 12:15:42.926   746   853 V FirebasePlugin: ______________________________Notification extras : Bundle[mParcelledData.dataSize=36]`

So it is always present but then right after, in the log we get

```
10-06 12:15:42.940 14365 14365 W Binder_5: type=1400 audit(0.0:2067): avc: denied { ioctl } for path="socket:[909661]" dev="sockfs" ino=909661 ioctlcmd=7704 scontext=u:r:surfaceflinger:s0 tcontext=u:r:surfaceflinger:s0 tclass=unix_stream_socket permissive=0

10-06 12:15:42.940 14365 14365 W Binder_5: type=1400 audit(0.0:2068): avc: denied { ioctl } for path="socket:[909661]" dev="sockfs" ino=909661 ioctlcmd=7704 scontext=u:r:surfaceflinger:s0 tcontext=u:r:surfaceflinger:s0 tclass=unix_stream_socket permissive=0

10-06 12:15:42.950  5258  5258 W Binder_E: type=1400 audit(0.0:2069): avc: denied { ioctl } for path="socket:[908177]" dev="sockfs" ino=908177 ioctlcmd=7704 scontext=u:r:system_server:s0 tcontext=u:r:system_server:s0 tclass=unix_stream_socket permissive=0

10-06 12:15:42.950  5258  5258 W Binder_E: type=1400 audit(0.0:2070): avc: denied { ioctl } for path="socket:[908177]" dev="sockfs" ino=908177 ioctlcmd=7704 scontext=u:r:system_server:s0 tcontext=u:r:system_server:s0 tclass=unix_stream_socket permissive=0

```

Not too sure at this point if it's related or what

**Comment by dpa99c** [comment]
You can't stringify the extras Bundle - you'll just get that meaningless class name summary - you need to step-through debug in Android Studio which can interpolate the class properties into a meaningful representation.

The other log warnings are completely unrelated.

**Comment by globules-io** [comment]
I don't think I can debug the app while it's in the background though...

**Comment by dpa99c** [comment]
> I don't think I can debug the app while it's in the background though...

You can't but you don't want to. You want to debug it when it launches in the foreground after a message arrives while it's not running or in the background. 

When I get some time I will be able to look into it.

**Comment by globules-io** [comment]
Ok thanks, I'll keep on debugging, java is not my forte. Thanks.

**Comment by globules-io** [comment]
So I traced it back the issue and extra does not contain `google.message_id` 

                   ```
    if (extras.containsKey("google.message_id")) {
                            extras.putString("messageType", "notification");
                            extras.putString("tap", "background");
                            notificationStack.add(extras);
                            Log.d(TAG, "Notification message found on init: " + extras.toString());
                        }

```

**Comment by globules-io** [comment]
If you have any idea how to fix, please let us know. As it is, right now, Firebase notifications as not reliable enough to consider using the data parameter if the data doesn't reach the app if it is running in the background...

**Comment by dpa99c** [comment]
I'll look into this when I have some time

**Comment by globules-io** [comment]
Thought I'd add all the possible workflows for notifications, if that can help you out.

- App is not running, notification is received, user taps notification, apps starts and data is passed to app : OK
- App is running in the foreground, data is passed to app : OK
- App is running in the background, notification is received, user swipes off notification, user brings back app to foreground: NOT PASSED
- App is running in the background, notification is received, user does not tap notification, user brings back app to foreground: NOT PASSED

That's all the possible cases I could think of.

**Comment by dpa99c** [comment]
> App is running in the background, notification is received, user swipes off notification, user brings back app to foreground: NOT PASSED

According to the [Firebase documentation](https://firebase.google.com/docs/cloud-messaging/android/receive), I don't believe this is possible on Android:


> __Messages with both notification and data payload, when received in the background__. In this case, the notification is delivered to the device’s system tray, and the data payload is delivered in the extras of the intent of your launcher Activity.

This means the data payload will only be delivered on app start when your launch intent of your main activity (which is `CordovaActivity` in a Cordova app) is invoked.
And this only happens at app startup.
So you'd have to wait until the app start (not resume) for this intent to be invoked and the extras bundle containing the data payload to be delivered.

Note: the launch intent is only invoked for messages that contain **both** notification **and** data payloads.

Data-only messages will be delivered to `onMessageReceived` if the app is running in the foreground **or** in the background but **never** if the app is not running/has been task-killed ([see here](https://github.com/firebase/quickstart-android/issues/368)).

**Comment by globules-io** [comment]
Ok thank you for you investigation. So from what I understand, the only way to have a full proof notification + data is to send 2 notifications. One with only title and body without data, and one with only data. Do you agree?
Thank you.

**Comment by dpa99c** [comment]
> Ok thank you for you investigation. So from what I understand, the only way to have a full proof notification + data is to send 2 notifications. One with only title and body without data, and one with only data. Do you agree?
> Thank you.

No I don't believe this is the case. 

"One with only title and body without data" is a "notification message". 
This means it will only be received by the app if it's running in the foreground **or** if in background/not running **and** user taps the system notification.

"one with only data" is a "data message" which will only be received in the app is running in the foreground/background but **not** if the app is not running.

Which means if the app is not running and the user didn't tap the system notification, you wouldn't receive either message.

According to the Firebase docs, the only way to guarantee that a message is delivered is to send "notification+data message" in which case:
- if the app is running in the foreground, both the notification payload and the data payload will be delivered to `onMessageReceived`
- if the app is running in the background **or** is not running
	- the notification payload will be delivered to the system tray
		- if the user taps the system notification, it will be delivered to `onMessageReceived`
	- the data payload will be delivered in the extras bundle of your app's launch intent
		- which means the next time your cold app starts (not resumes); this will invoke this plugin's `pluginInitialize` method which should read the data from the launch intent

**Comment by globules-io** [comment]
Yes right, thank you but let me say this again

- If my app is running in the background and a message with notification and data is sent, and the user discards it or switches back to the app without tapping it, the data is never passed, unless the app is restarted.

The only way to have data passed, from my understanding, would be to send 2 messages, one with a notification and data, and another one with data only for make sure it reaches the app if it is running in the background, in case the end user discards the notification or does not tap it and switches back to the app.

Then it's about testing if the 2nd message with data only was already received by the app IF the end-user taps the notification on the first message.

Do you see what I mean? If an app running in the background only get data with a message with data only, this means that sending a message with notification + data is not full proof. It needs to be duplicated.

**Comment by dpa99c** [comment]
> * If my app is running in the background and a message with notification and data is sent, and the user discards it or switches back to the app without tapping it, the data is never passed, unless the app is restarted.

"unless the app is restarted" is the key phrase. This is (according to the Firebase docs) the only **guaranteed** way that your app will receive the message (as a data payload).

> The only way to have data passed, from my understanding, would be to send 2 messages, one with a notification and data, and another one with data only for make sure it reaches the app if it is running in the background, in case the end user discards the notification or does not tap it and switches back to the app.

> Then it's about testing if the 2nd message with data only was already received by the app IF the end-user taps the notification on the first message.

The data payload of a data-only message is **never** received if your app is not running.
 
> Do you see what I mean? If an app running in the background only get data with a message with data only, this means that sending a message with notification + data is not full proof. It needs to be duplicated.

According to the Firebase docs, a "notification+data" message is the only **guaranteed** way to deliver a data payload to the app. However, this comes with the caveat that if the app is backgrounded/not running, the data payload will not be delivered until the next launch intent (app start). But (at least according to the Firebase docs) it will **always** be delivered, regardless of your app's running state. So duplication with a second data-only message in unnecessary since (if we believe the Firebase docs) the data payload will **eventually** always be delivered (unless the user uninstalls the app or never starts it again).

**Comment by globules-io** [comment]
Yes I agree with you but here's the thing. Apps usually don't really care about notifications at start because most probably they will do a server fetch for the latest data anyway. Now I'm not arguing with what you are saying, I agree, but it does not cover the background case. I will do a little more testing with the workflows and will report, just to share my findings, and will close this issue then.
Thank you for your time.

**Comment by dpa99c** [comment]
I've looked into this a bit more between other things:

According to [this Firebase documentation](https://firebase.google.com/docs/cloud-messaging/concept-options#notification-messages-with-optional-data-payload):

> When in the background, apps receive the notification payload in the notification tray, and **only handle the data payload when the user taps on the notification**.

This is contradictory to the [other Firebase documentation](https://firebase.google.com/docs/cloud-messaging/android/receive#handling_messages) which seems to imply the data payload will be preserved ("stacked up") and sent to the launch intent when the application is next manually launched.


So I decided to do some empirical testing with the currently plugin implementation:

When a notification+data message arrives while the app is not running, when I tap the system notification, I can see the launch intent being invoked with an extras bundle containing the data payload.

But if I don't tap the system notification and I manually launch the app, the extras bundle is empty.

TL;DR: contrary to my first understanding, the data payload for a notification+data message is **only delivered if/when the user taps on the system notification**.
It the user does not tap, the app never receives the data payload.

Therefore the current implementation of this plugin complies with what the Firebase documentation states is possible.
I don't believe anything more can be done from a code perspective to improve the current behaviour but I will add to the documentation to make it clear exactly what and what is not possible with the various message payloads.

**Comment by dpa99c** [comment]
I believe the documentation added by the above commit clarifies the behaviour as much as is possible.

**Comment by globules-io** [comment]
Going to leave this final note to whoever read the issues and faces the same problem.
As @dpa99c stated, and thanks for his time and efforts, the plugin behaves properly. 

Here is a work around to ensure that data is always delivered in every scenario:
- If you are sending both notification and data in your message, send them separately because data alone always reaches the app (foreground or background) while notification+data is mainly based on the end user behavior.
- If the app is running, you can ignore the tap event because data is passed anyway with the data broadcast only.
- Broadcast the data message first, then the notification.
